p2p/src/network/pubsub/p2p_network_pubsub_reducer.rs

```diff
@@ -140,56 +203,71 @@ impl P2pNetworkPubsubState {
                 }
                 for iwant in &control.iwant {
                     for msg_id in &iwant.message_ids {
-                        if let Some(msg) = self.mcache.map.get(msg_id) {
-                            if let Some(client) = self.clients.get_mut(peer_id) {
+                        if let Some(msg) = pubsub_state.mcache.map.get(msg_id) {
+                            if let Some(client) = pubsub_state.clients.get_mut(peer_id) {
                                 client.message.publish.push(msg.clone());
                             }
                         }
                     }
                 }
+
                 for ihave in &control.ihave {
                     let message_ids = ihave
                         .message_ids
                         .iter()
-                        .filter(|msg_id| !self.mcache.map.contains_key(*msg_id))
+                        .filter(|msg_id| !pubsub_state.mcache.map.contains_key(*msg_id))
                         .cloned()
                         .collect();
-                    if let Some(client) = self.clients.get_mut(peer_id) {
+                    if let Some(client) = pubsub_state.clients.get_mut(peer_id) {
                         let ctr = client.message.control.get_or_insert_with(Default::default);
                         ctr.iwant.push(pb::ControlIWant { message_ids })
                     }
                 }
+
+                let dispatcher: &mut Dispatcher<Action, State> = state_context.into_dispatcher();
+                dispatcher.push(P2pNetworkPubsubEffectfulAction::IncomingData {
+                    peer_id: *peer_id,
+                    seen_limit: *seen_limit,
+                });
+
+                Ok(())
             }
             P2pNetworkPubsubAction::IncomingMessage {
```

Comment:
the body for this action handler is quite big with many things going on, would it make sense to reorganize it a bit to use some helpers?

Reply:
I have moved this to: `reduce_incoming_message` I have also moved reducer for `P2pNetworkPubsubAction::IncomingData` to `reduce_incoming_data`, since that reducer is also big

Reply:
thank you, much better now